python-threatexchange/threatexchange/fetcher/simple/state.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+
+from collections import defaultdict
+from dataclasses import dataclass, field
+import logging
+import typing as t
+from threatexchange.fetcher.fetch_api import SignalExchangeAPI
+
+from threatexchange.signal_type.signal_base import SignalType
+from threatexchange.fetcher import fetch_state
+from threatexchange.fetcher.collab_config import CollaborationConfigBase
+
+
+@dataclass
+class SimpleFetchedSignalMetadata(fetch_state.FetchedSignalMetadata):
+    """
+    Simple dataclass for fetched data.
+
+    Merge by addition rather than replacement.
+    """
+
+    opinions: t.List[fetch_state.SignalOpinion] = field(default_factory=list)
+
+    def get_as_opinions(self) -> t.List[fetch_state.SignalOpinion]:
+        return self.opinions
+
+    @classmethod
+    def merge(
+        cls, older: "SimpleFetchedSignalMetadata", newer: "SimpleFetchedSignalMetadata"
+    ) -> "SimpleFetchedSignalMetadata":
+        if not older.opinions:
+            return newer
+
+        by_owner = {o.owner: o for o in newer.opinions}
+        return cls([by_owner.get(o.owner, o) for o in older.opinions])
+
+    @classmethod
+    def get_trivial(cls):
+        return cls([fetch_state.SignalOpinion.get_trivial()])
+
+
+@dataclass
+class SimpleFetchDelta(fetch_state.FetchDeltaWithUpdateStream):
+    """
+    Simple class for deltas.
+
+    If the record is set to None, this indicates the record should be
+    deleted if it exists.
+    """
+
+    updates: t.Mapping[t.Tuple[str, str], t.Optional[fetch_state.FetchedSignalMetadata]]
+    checkpoint: fetch_state.FetchCheckpointBase
+    done: bool  # powers has_more
+
+    def record_count(self) -> int:
+        return len(self.updates)
+
+    def next_checkpoint(self) -> fetch_state.FetchCheckpointBase:
+        return self.checkpoint
+
+    def has_more(self) -> bool:
+        return not self.done
+
+    def get_as_update_dict(
+        self,
+    ) -> t.Mapping[t.Tuple[str, str], t.Optional[fetch_state.FetchedSignalMetadata]]:
+        return self.updates
+
+
+@dataclass
+class _StateTracker:
+    updates_by_type: t.Dict[str, t.Dict[str, fetch_state.FetchedSignalMetadata]]
+    checkpoint: t.Optional[fetch_state.FetchCheckpointBase]
+    dirty: bool = False
+
+    def merge(self, newer: fetch_state.FetchDeltaWithUpdateStream) -> None:
+        updates = newer.get_as_update_dict()
+        if not updates:
+            return
+        newer_by_type: t.DefaultDict[
+            str, t.List[t.Tuple[str, t.Optional[fetch_state.FetchedSignalMetadata]]]
+        ] = defaultdict(list)
+        for (stype, signal_str), record in updates.items():
+            newer_by_type[stype].append((signal_str, record))
+
+        for n_type, n_updates in newer_by_type.items():
```

Comment:
Could definitely use a test case. Add as a card to project if not already done.

Reply:
Done